hazelcast/src/main/java/com/hazelcast/spi/impl/operationexecutor/OperationExecutor.java

```diff
@@ -95,7 +99,15 @@
      * @throws java.lang.NullPointerException if op is null.
      * @throws IllegalThreadStateException    if the operation is not allowed to be run on the calling thread.
      */
-    void runOperationOnCallingThread(Operation op);
+    void runOnCallingThread(Operation op);
+
+    /**
+     * Tries to run the operation on the calling thread if possible. Otherwise offload it to a different thread.
```

Comment:
sure. I understand performance implications of off-loading a task to a different thread. my understanding is that for some operations this method will block until the operation is finished, but for other (off-loaded) operations it will return immediately. is it so? it should be probably explicitly stated in the contract. 

maybe just a word of warning that in the case of off-loading it's not waiting for the operation to be finished.